typed-protocols/src/Network/TypedProtocol/Pipelined.hs

```diff
@@ -44,16 +45,21 @@ data PeerSender ps (pr :: PeerRole) (st :: ps) (n :: Outstanding) c m a where
                  -> a
                  -> PeerSender ps pr st Z c m a
 
-  -- | A normal non-pipelined 'Yield'. Note that this can only be used for
-  -- messages that keep agency (or terminal messages), not for ones that
-  -- give away agency, since the 'PeerSender' has no equivalent of 'Await'.
-  -- The only way to await is via a pipelined yield: 'SenderPipeline'.
+  -- | A normal non-pipelined 'Yield'.
   --
   SenderYield    :: !(WeHaveAgency pr st)
                  -> Message ps st st'
                  -> PeerSender   ps pr (st' :: ps) n c m a
                  -> PeerSender   ps pr (st  :: ps) n c m a
 
+  -- | A normal non-pipelined 'Await'.  We only allow await if there are no
+  -- pipelined requests.
+  --
+  SenderAwait    :: !(TheyHaveAgency pr st)
+                 -> (forall (st' :: ps). Message ps st st'
+                      -> PeerSender ps pr (st' :: ps) Z c m a)
+                 -> PeerSender ps pr (st :: ps) Z c m a
+
```

Comment:
What is the motivation for this? It does make the peer driver more tricky because of the sharing of the trailing data between two threads. This can be simulated by doing a send with an immediate collect can't it? Can we just do this as a library function rather than a primitive?

It'd be something like
```
senderSendAndAwait stok msg stok' =
  SenderPipeline stok msg
    (ReceiverAwait stok' $ \msg' -> ReceiverDone msg')
```

Reply:
In `tx-submission` we
* ask for hashes
* recieve them
* do pipelineing when getting each transactions

The current `PeerPipelined` has a single `c` parameter which has to be shared across all pipelined parts which is not very elegant. `senderSendAndAwait` would need to use a sum type `Either [hash] c` and `Collect` right after `senderSendAndAwait` would need to hope that its `Left` rather than `Right`.  Allowing `SenderAwait` at the top level avoids this.

Another approach would be to make `c` depend on the final state of a pipelined transition sequence.

Reply:
But don't we usually want to pipeline asking for hashes too? It's only a special case when we have no hashes to ask for that we need to wait. Hmm, perhaps that suggests we need two kinds of "ask for hashes" messages, one where it blocks until there are some, and another that's non-blocking and can return none.

Reply:
@dcoutts that's a good point, we need both pipelined and non pipelined `MsgGetHashes`.

Reply:
I saved the `SenderAwait` in a local branch, and removed it from this PR.  I have an impression that, we could avoid the complexity with the trailing bytes for `SenderAwait` by removing concurrency as we do with the `Queue`, we'd just need a more efficient queue.